testng/src/main/java/cucumber/api/testng/TestNGCucumberRunner.java

```diff
@@ -97,4 +102,25 @@ public void finish() {
         }
     }
 
+    /**
+     * @return returns the cucumber scenarios as a two dimensional array of {@link PickleEventWrapper}
+     * scenarios combined with their {@link CucumberFeatureWrapper} feature.
+     */
+    public Object[][] provideScenarios() {
+        try {
+            List<Object[]> scenarios = new ArrayList<Object[]>();
+
+            List<CucumberFeature> features = getFeatures();
+            for (CucumberFeature feature : features) {
+                List<PickleEvent> pickles = runtime.compileFeature(feature);
+
+                for (PickleEvent pickle : pickles) {
+                    scenarios.add(new Object[]{new PickleEventWrapper(pickle), new CucumberFeatureWrapperImpl(feature)});
```

Comment:
It is here that `runtime.matchesFilters(pickle)` should be used to select the pickles to run.